app/src/main/java/ch/epfl/sdp/musiconnect/MapsActivity.java

```diff
@@ -305,7 +385,114 @@ public void onInfoWindowClick(Marker marker) {
         }
     }
 
+    protected boolean checkConnection() {
+        ConnectivityManager connectivityManager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
+        if (connectivityManager.getNetworkInfo(ConnectivityManager.TYPE_MOBILE).getState() == NetworkInfo.State.CONNECTED ||
+                connectivityManager.getNetworkInfo(ConnectivityManager.TYPE_WIFI).getState() == NetworkInfo.State.CONNECTED) {
+            //we are connected to a network
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    protected boolean checkLocationServices(){
+        LocationManager lm = (LocationManager)MapsActivity.this.getSystemService(Context.LOCATION_SERVICE);
+        boolean gps_enabled = false;
+        boolean network_enabled = false;
+
+        try {
+            gps_enabled = lm.isProviderEnabled(LocationManager.GPS_PROVIDER);
+        } catch(Exception ex) {}
+
+        try {
+            network_enabled = lm.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
+        } catch(Exception ex) {}
+
+        if(!gps_enabled && !network_enabled) {
+            return false;
+        } else{
+            return true;
+        }
+
+    }
+
+
+
+    private void saveUsersToCache() {
+        FileOutputStream fos = null;
+        String toCache = sdf.format(timeLastUpdt) + "\n";
+        for(Musician p:allUsers){
+            String birthdate = p.getBirthday().getYear() + "/" + p.getBirthday().getMonth() + "/" + p.getBirthday().getDate();
+            toCache = toCache + p.getFirstName() + "," + p.getLastName() + "," + p.getUserName() + ","
```

Comment:
All these concatenations are ugly and not very efficient performance-wise, maybe use a String.format(). Also, as discussed during the meeting, this is probably not what you want to do for a cache.